runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,219 @@
+                streamManager.unsubscribe(listener);
+                return;
+            }
+
+            while (!isShutdown()) {
+                try {
+                    while (!txnStream.hasNext()) {
+                        log.debug("{}::run() all caught up. Taking a 50ms nap", listener.toString());
+                        int DEFAULT_TIMEOUT_IN_MILLI = 50;
+                        TimeUnit.MILLISECONDS.sleep(DEFAULT_TIMEOUT_IN_MILLI);
+                    }
+                    ILogData logData = txnStream.nextUpTo(Address.MAX);
+                    MultiObjectSMREntry multiObjSMREntry = (MultiObjectSMREntry) logData.getPayload(runtime);
+                    long epoch = logData.getEpoch();
+                    Map<TableSchema, List<CorfuStreamEntry>> entries = new HashMap<>();
+                    // first only filter by the stream IDs of interest that are present in this logData
+                    logData.getStreams().stream().filter(tablesOfInterest::containsKey)
+                            .forEach(streamId -> entries.put(tablesOfInterest.get(streamId),
+                            // Only extract the list of updates per stream as a list
+                            multiObjSMREntry.getSMRUpdates(streamId).stream().map(smrEntry ->
+                                    CorfuStreamEntry.fromSMREntry(smrEntry,
+                                    epoch,
+                                    tablesOfInterest.get(streamId).getKeyClass(),
+                                    tablesOfInterest.get(streamId).getPayloadClass(),
+                                    tablesOfInterest.get(streamId).getMetadataClass())
+                            ).collect(Collectors.toList())));
+
+                    if (!entries.isEmpty()) {
+                        CorfuStreamEntries callbackResult = new CorfuStreamEntries(entries);
+                        long onNextStart = System.nanoTime();
+                        listener.onNext(callbackResult);
+                        long onNextEnd = System.nanoTime();
+                        log.debug("{}::onNext took {} ms", listener.toString(),
+                                TimeUnit.NANOSECONDS.toMillis(onNextEnd - onNextStart));
+                    }
+                } catch (Throwable throwable) {
+                    log.warn("{}::onError: {}", listener.toString(), throwable.toString());
+                    listener.onError(throwable);
+                    streamManager.unsubscribe(listener);
+                    return;
+                }
+            }
+        }
+
+        public SubscriptionThread(@Nonnull StreamManager streamManager,
+                                  @Nonnull StreamListener listener,
+                                  @Nonnull String namespace,
+                                  @Nonnull List<TableSchema> tableSchemas,
+                                  long startAddress) {
+            super("CorfuSubscriber: "+namespace+":"+listener.toString());
```

Comment:
nit: spaces around '+'

Reply:
also isnt it more clear this way? "corfusubscriber: listener.toString() + namesapace"?

Reply:
yes it is, fixed, thanks